# Análisis Exploratorio del Titanic

**Curso:** Programación II — Ciencia de Datos  
**Universidad:** Externado de Colombia  
**Integrantes:** Jacobo Beltrán y Helen Rueda

## Introducción

El dataset del Titanic contiene información sobre los pasajeros del famoso
barco que naufragó en 1912. El objetivo de este análisis es explorar los
datos, entender las variables y responder preguntas sobre los pasajeros
y sus probabilidades de supervivencia.

Los datos provienen de Kaggle y están divididos en dos conjuntos:
- **train.csv**: 891 pasajeros con información de supervivencia
- **test.csv**: 418 pasajeros sin información de supervivencia

## 1. Importación de librerías y carga de datos

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

plt.style.use('seaborn-v0_8')
sns.set_palette('husl')

print("Librerias importadas correctamente")
print(f"Version de Pandas: {pd.__version__}")
print(f"Version de NumPy: {np.__version__}")

Librerias importadas correctamente
Version de Pandas: 3.0.3
Version de NumPy: 2.4.6


In [3]:
import os

base = os.path.dirname(os.path.abspath('proyecto_titanic.ipynb'))
ruta_train = os.path.join(base, '..', 'data', 'train.csv')
ruta_test = os.path.join(base, '..', 'data', 'test.csv')

df_train = pd.read_csv(ruta_train)
df_test = pd.read_csv(ruta_test)

print("Datasets cargados correctamente")
print(f"Filas en train: {df_train.shape[0]} | Columnas en train: {df_train.shape[1]}")
print(f"Filas en test:  {df_test.shape[0]} | Columnas en test:  {df_test.shape[1]}")

Datasets cargados correctamente
Filas en train: 891 | Columnas en train: 12
Filas en test:  418 | Columnas en test:  11


In [4]:
print("=== Primeras filas de train.csv ===")
df_train.head()

=== Primeras filas de train.csv ===


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [5]:
print("=== Primeras filas de test.csv ===")
df_test.head()

=== Primeras filas de test.csv ===


,PassengerId,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,892,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,NaN,Q
1,893,3,"Wilkes, Mrs. James (Ellen Needs)",female,47.0,1,0,363272,7.0000,NaN,S
2,894,2,"Myles, Mr. Thomas Francis",male,62.0,0,0,240276,9.6875,NaN,Q
3,895,3,"Wirz, Mr. Albert",male,27.0,0,0,315154,8.6625,NaN,S
4,896,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22.0,1,1,3101298,12.2875,NaN,S


## 2. Descripción de las variables

El dataset del Titanic contiene las siguientes variables:

| Variable | Tipo | Descripción |
|----------|------|-------------|
| PassengerId | Numérica | Identificador único de cada pasajero |
| Survived | Categórica | Supervivencia (0 = No, 1 = Sí). Solo en train |
| Pclass | Categórica | Clase del tiquete (1 = Primera, 2 = Segunda, 3 = Tercera) |
| Name | Texto | Nombre completo del pasajero |
| Sex | Categórica | Género del pasajero (male/female) |
| Age | Numérica | Edad en años. Puede tener valores faltantes |
| SibSp | Numérica | Número de hermanos o cónyuges a bordo |
| Parch | Numérica | Número de padres o hijos a bordo |
| Ticket | Texto | Número del tiquete |
| Fare | Numérica | Tarifa pagada por el pasajero |
| Cabin | Texto | Número de cabina. Tiene muchos valores faltantes |
| Embarked | Categórica | Puerto de embarque (C = Cherburgo, Q = Queenstown, S = Southampton) |

In [6]:
print("=== Tipos de datos — train ===")
print(df_train.dtypes)
print(f"\nTotal de variables: {df_train.shape[1]}")

=== Tipos de datos — train ===
PassengerId      int64
Survived         int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Cabin              str
Embarked           str
dtype: object

Total de variables: 12


In [7]:
print("=== Tipos de datos — test ===")
print(df_test.dtypes)
print(f"\nTotal de variables: {df_test.shape[1]}")
print("\nNota: test no tiene la variable Survived")

=== Tipos de datos — test ===
PassengerId      int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Cabin              str
Embarked           str
dtype: object

Total de variables: 11

Nota: test no tiene la variable Survived


## 3. Estadísticas básicas y comparación entre datasets

In [8]:
print("=== Estadísticas básicas — train ===")
df_train.describe()

=== Estadísticas básicas — train ===


,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [9]:
print("=== Estadísticas básicas — test ===")
df_test.describe()

=== Estadísticas básicas — test ===


,PassengerId,Pclass,Age,SibSp,Parch,Fare
count,418.000000,418.000000,332.000000,418.000000,418.000000,417.000000
mean,1100.500000,2.265550,30.272590,0.447368,0.392344,35.627188
std,120.810458,0.841838,14.181209,0.896760,0.981429,55.907576
min,892.000000,1.000000,0.170000,0.000000,0.000000,0.000000
25%,996.250000,1.000000,21.000000,0.000000,0.000000,7.895800
50%,1100.500000,3.000000,27.000000,0.000000,0.000000,14.454200
75%,1204.750000,3.000000,39.000000,1.000000,0.000000,31.500000
max,1309.000000,3.000000,76.000000,8.000000,9.000000,512.329200


In [10]:
print("=== Comparación entre datasets ===\n")
print(f"{'':20} {'TRAIN':>10} {'TEST':>10}")
print(f"{'Filas':20} {df_train.shape[0]:>10} {df_test.shape[0]:>10}")
print(f"{'Columnas':20} {df_train.shape[1]:>10} {df_test.shape[1]:>10}")
print(f"{'Edad promedio':20} {df_train['Age'].mean():>10.2f} {df_test['Age'].mean():>10.2f}")
print(f"{'Tarifa promedio':20} {df_train['Fare'].mean():>10.2f} {df_test['Fare'].mean():>10.2f}")
print(f"{'% mujeres':20} {(df_train['Sex']=='female').mean()*100:>10.2f} {(df_test['Sex']=='female').mean()*100:>10.2f}")
print(f"\nNota: train tiene la columna Survived, test no la tiene")

=== Comparación entre datasets ===

                          TRAIN       TEST
Filas                       891        418
Columnas                     12         11
Edad promedio             29.70      30.27
Tarifa promedio           32.20      35.63
% mujeres                 35.24      36.36

Nota: train tiene la columna Survived, test no la tiene


### Comparación entre datasets

Al comparar train y test observamos que:
- **train** tiene 891 pasajeros y **test** tiene 418
- La edad promedio y la tarifa promedio son similares entre ambos datasets
- La proporción de mujeres es comparable entre los dos conjuntos
- Esto sugiere que ambos datasets son representativos de la misma población

## 4. Nueva variable: Familiares

In [11]:
df_train['Familiares'] = df_train['SibSp'] + df_train['Parch']
df_test['Familiares'] = df_test['SibSp'] + df_test['Parch']

print("Variable Familiares creada correctamente")
print(f"\nDistribucion en train:")
print(df_train['Familiares'].value_counts().sort_index())

Variable Familiares creada correctamente

Distribucion en train:
Familiares
0     537
1     161
2     102
3      29
4      15
5      22
6      12
7       6
10      7
Name: count, dtype: int64


### Análisis de la variable Familiares

La variable **Familiares** representa el número total de familiares
que viajaban con cada pasajero a bordo del Titanic.

**Observaciones:**
- La mayoría de los pasajeros viajaban solos (Familiares = 0)
- Los pasajeros con familias pequeñas (1 a 3 familiares) son el
  segundo grupo más grande
- Muy pocos pasajeros viajaban con familias grandes (4 o más)
- Esta variable será útil para analizar si viajar acompañado
  influyó en las probabilidades de supervivencia

## 5. Unión vertical de los datasets

In [12]:
df_train['source'] = 'train'
df_test['source'] = 'test'

df_test['Survived'] = None

print("Columnas en train:", df_train.columns.tolist())
print("Columnas en test: ", df_test.columns.tolist())

Columnas en train: ['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked', 'Familiares', 'source']
Columnas en test:  ['PassengerId', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked', 'Familiares', 'source', 'Survived']


In [13]:
df_total = pd.concat([df_train, df_test], ignore_index=True)

print("Dataset combinado creado correctamente")
print(f"Filas totales: {df_total.shape[0]}")
print(f"Columnas:      {df_total.shape[1]}")
print(f"\nDistribucion por fuente:")
print(df_total['source'].value_counts())

Dataset combinado creado correctamente
Filas totales: 1309
Columnas:      14

Distribucion por fuente:
source
train    891
test     418
Name: count, dtype: int64


### Decisión sobre la columna Survived en test

El dataset **test** no contiene la variable `Survived` porque
en la competencia de Kaggle es la variable que se debe predecir.

Para poder unir ambos datasets verticalmente sin perder información,
agregamos la columna `Survived` al dataset test con valores `NaN`,
lo que indica que no se conoce la supervivencia de esos pasajeros.

El dataset combinado tiene **1309 pasajeros** en total:
- 891 de train (con información de supervivencia)
- 418 de test (sin información de supervivencia)

## 6. Revision de estadisticas del dataset combinado

In [14]:
# Estadisticas generales del dataset combinado
print("=== Estadisticas del dataset combinado ===")
print(df_total.describe())

print(f"\nForma del dataset: {df_total.shape[0]} filas x {df_total.shape[1]} columnas")
print(f"\nDistribucion por fuente:")
print(df_total['source'].value_counts())

=== Estadisticas del dataset combinado ===
       PassengerId       Pclass          Age        SibSp        Parch  \
count  1309.000000  1309.000000  1046.000000  1309.000000  1309.000000   
mean    655.000000     2.294882    29.881138     0.498854     0.385027   
std     378.020061     0.837836    14.413493     1.041658     0.865560   
min       1.000000     1.000000     0.170000     0.000000     0.000000   
25%     328.000000     2.000000    21.000000     0.000000     0.000000   
50%     655.000000     3.000000    28.000000     0.000000     0.000000   
75%     982.000000     3.000000    39.000000     1.000000     0.000000   
max    1309.000000     3.000000    80.000000     8.000000     9.000000   

              Fare   Familiares  
count  1308.000000  1309.000000  
mean     33.295479     0.883881  
std      51.758668     1.583639  
min       0.000000     0.000000  
25%       7.895800     0.000000  
50%      14.454200     0.000000  
75%      31.275000     1.000000  
max     512.329200

### 7. Variables con datos faltantes y su significado

In [15]:
# Conteo de valores faltantes por columna en el dataset combinado
faltantes = df_total.isnull().sum()
faltantes_pct = (df_total.isnull().sum() / len(df_total)) * 100

resumen_faltantes = pd.DataFrame({
    'Faltantes': faltantes,
    'Porcentaje': faltantes_pct.round(2)
})

resumen_faltantes = resumen_faltantes[resumen_faltantes['Faltantes'] > 0].sort_values('Faltantes', ascending=False)

print("=== Variables con datos faltantes ===")
print(resumen_faltantes)

=== Variables con datos faltantes ===
          Faltantes  Porcentaje
Cabin          1014       77.46
Survived        418       31.93
Age             263       20.09
Embarked          2        0.15
Fare              1        0.08


### Interpretacion de los valores faltantes 

- **Survived**: 418 faltantes esperados, corresponden al dataset test
- **Cabin**: mas del 77% de valores faltantes, es una variable poco util para el analisis
- **Age**: aproximadamente 20% de faltantes, se debe considerar al analizar edad
- **Fare**: 1 faltante en test, valor puntual sin mayor impacto
- **Embarked**: 2 faltantes en train, valor puntual sin mayor impacto

## 8. Preguntas concretas sobre los pasajeros

### 8a. Cual es la edad promedio de los pasajeros

In [16]:
 # Edad promedio de los pasajeros en el dataset combinado
edad_promedio = df_total['Age'].mean()
edad_mediana = df_total['Age'].median()
edad_min = df_total['Age'].min()
edad_max = df_total['Age'].max()
faltantes_edad = df_total['Age'].isnull().sum()

print("=== Edad de los pasajeros ===")
print(f"Edad promedio:  {edad_promedio:.2f} anos")
print(f"Edad mediana:   {edad_mediana:.2f} anos")
print(f"Edad minima:    {edad_min:.0f} anos")
print(f"Edad maxima:    {edad_max:.0f} anos")
print(f"Valores faltantes en Age: {faltantes_edad}")

=== Edad de los pasajeros ===
Edad promedio:  29.88 anos
Edad mediana:   28.00 anos
Edad minima:    0 anos
Edad maxima:    80 anos
Valores faltantes en Age: 263


### Observacion

El promedio y la mediana de edad son similares, lo que indica una distribucion
relativamente simetrica. Sin embargo, hay valores faltantes en Age que se deben
tener en cuenta al interpretar estos resultados.

### 8c. Tarifa promedio de pasajeros de primera clase

In [17]:
# Tarifa promedio por clase de pasajero
tarifa_por_clase = df_total.groupby('Pclass')['Fare'].mean().round(2)

print("=== Tarifa promedio por clase ===")
for clase, tarifa in tarifa_por_clase.items():
    print(f"Clase {clase}: ${tarifa}")

tarifa_primera = df_total[df_total['Pclass'] == 1]['Fare'].mean()
print(f"\nLa tarifa promedio de primera clase fue ${tarifa_primera:.2f}")
print(f"Es {(tarifa_primera / df_total[df_total['Pclass'] == 3]['Fare'].mean()):.1f} veces mas cara que tercera clase")

=== Tarifa promedio por clase ===
Clase 1: $87.51
Clase 2: $21.18
Clase 3: $13.3

La tarifa promedio de primera clase fue $87.51
Es 6.6 veces mas cara que tercera clase


### Observacion

La tarifa de primera clase es significativamente mas alta que la de segunda
y tercera clase, lo que refleja las grandes diferencias economicas entre los
grupos de pasajeros a bordo del Titanic.

### 8e. Edad mas joven y mas vieja

In [18]:
# Pasajero mas joven y mas viejo
edad_min = df_total['Age'].min()
edad_max = df_total['Age'].max()

pasajero_joven = df_total[df_total['Age'] == edad_min][['Name', 'Age', 'Pclass', 'Sex', 'source']].head(1)
pasajero_viejo = df_total[df_total['Age'] == edad_max][['Name', 'Age', 'Pclass', 'Sex', 'source']].head(1)

print("=== Pasajero mas joven ===")
print(pasajero_joven.to_string(index=False))

print("\n=== Pasajero mas viejo ===")
print(pasajero_viejo.to_string(index=False))

=== Pasajero mas joven ===
                                   Name  Age  Pclass    Sex source
Dean, Miss. Elizabeth Gladys Millvina"" 0.17       3 female   test

=== Pasajero mas viejo ===
                                Name  Age  Pclass  Sex source
Barkworth, Mr. Algernon Henry Wilson 80.0       1 male  train


### Observacion

El pasajero mas joven era un bebe, lo que muestra que habia familias completas
a bordo incluyendo ninos de muy corta edad. El pasajero mas viejo tenia mas de
70 anos, lo que era poco comun para un viaje transatlantico en esa epoca.

### 8f. Cuantos pasajeros viajaron desde cada puerto

In [19]:
# Distribucion de pasajeros por puerto de embarque
puertos = df_total['Embarked'].value_counts()
puertos_pct = (df_total['Embarked'].value_counts() / df_total['Embarked'].count() * 100).round(2)

print("=== Pasajeros por puerto de embarque ===")
print(f"Southampton (S): {puertos.get('S', 0)} pasajeros ({puertos_pct.get('S', 0)}%)")
print(f"Cherbourg   (C): {puertos.get('C', 0)} pasajeros ({puertos_pct.get('C', 0)}%)")
print(f"Queenstown  (Q): {puertos.get('Q', 0)} pasajeros ({puertos_pct.get('Q', 0)}%)")
print(f"\nValores faltantes en Embarked: {df_total['Embarked'].isnull().sum()}")

=== Pasajeros por puerto de embarque ===
Southampton (S): 914 pasajeros (69.93%)
Cherbourg   (C): 270 pasajeros (20.66%)
Queenstown  (Q): 123 pasajeros (9.41%)

Valores faltantes en Embarked: 2


### Observacion

La gran mayoria de los pasajeros embarco en Southampton, que era el puerto
principal de salida del Titanic. Cherbourg y Queenstown fueron paradas
intermedias con menos pasajeros.

### 8g. Cuantos pasajeros viajaron solos y cuantos con familiares

In [20]:
# Verificamos si Familiares existe en df_total, si no la creamos
if 'Familiares' not in df_total.columns:
    df_total['Familiares'] = df_total['SibSp'] + df_total['Parch']

# Clasificacion de pasajeros segun si viajaban solos o acompanados
solos = df_total[df_total['Familiares'] == 0].shape[0]
acompanados = df_total[df_total['Familiares'] > 0].shape[0]
total = df_total.shape[0]

print("=== Pasajeros solos vs acompanados ===")
print(f"Viajaban solos:        {solos} pasajeros ({solos/total*100:.2f}%)")
print(f"Viajaban acompanados:  {acompanados} pasajeros ({acompanados/total*100:.2f}%)")
print(f"Total:                 {total} pasajeros")

=== Pasajeros solos vs acompanados ===
Viajaban solos:        790 pasajeros (60.35%)
Viajaban acompanados:  519 pasajeros (39.65%)
Total:                 1309 pasajeros


### Observacion

La mayoria de los pasajeros viajaba solo, lo que sugiere que muchos eran
emigrantes o viajeros de negocios. Los pasajeros acompanados representan
una proporcion significativa, lo que indica que tambien habia muchas familias
a bordo.

### 9a. Sobrevivieron mas mujeres que hombres proporcionalmente

Creemos que si, las mujeres tuvieron mayor tasa de supervivencia que
los hombres porque en la evacuacion del Titanic se priorizo primero
a mujeres y ninos.

In [21]:
# Tasa de supervivencia por sexo usando df_train
supervivencia_sexo = df_train.groupby('Sex')['Survived'].agg(['sum', 'count', 'mean'])
supervivencia_sexo.columns = ['Sobrevivieron', 'Total', 'Tasa']
supervivencia_sexo['Tasa'] = (supervivencia_sexo['Tasa'] * 100).round(2)

print("=== Supervivencia por sexo ===")
print(supervivencia_sexo.to_string())

=== Supervivencia por sexo ===
        Sobrevivieron  Total   Tasa
Sex                                
female            233    314  74.20
male              109    577  18.89


### Observacion

Los datos confirman que las mujeres tuvieron una tasa de supervivencia
significativamente mayor que los hombres. Esto es consistente con la
politica de evacuacion del Titanic que priorizaba mujeres y ninos.

### 9c. Pasajeros mayores de 50 o menores de 10 sobrevivieron mas proporcionalmente

Creemos que los menores de 10 si tuvieron mayor tasa de supervivencia
por la politica de evacuacion. Sin embargo, los mayores de 50 posiblemente
no, ya que la edad avanzada podia dificultar el proceso de evacuacion.

In [22]:
# Comparacion de supervivencia por grupo etario extremo vs resto
grupo_extremo = df_train[(df_train['Age'] < 10) | (df_train['Age'] > 50)]
grupo_resto = df_train[(df_train['Age'] >= 10) & (df_train['Age'] <= 50)]

tasa_extremo = grupo_extremo['Survived'].mean() * 100
tasa_resto = grupo_resto['Survived'].mean() * 100

print("=== Supervivencia por grupo etario ===")
print(f"Menores de 10 o mayores de 50:")
print(f"  Total:         {len(grupo_extremo)}")
print(f"  Sobrevivieron: {grupo_extremo['Survived'].sum()}")
print(f"  Tasa:          {tasa_extremo:.2f}%")

print(f"\nResto de pasajeros (10 a 50 anos):")
print(f"  Total:         {len(grupo_resto)}")
print(f"  Sobrevivieron: {grupo_resto['Survived'].sum()}")
print(f"  Tasa:          {tasa_resto:.2f}%")

=== Supervivencia por grupo etario ===
Menores de 10 o mayores de 50:
  Total:         126
  Sobrevivieron: 60
  Tasa:          47.62%

Resto de pasajeros (10 a 50 anos):
  Total:         588
  Sobrevivieron: 230
  Tasa:          39.12%


### Observacion

Los datos muestran que los pasajeros en edades extremas tuvieron una
tasa de supervivencia ligeramente diferente al resto. Los menores de
10 anos se beneficiaron de la politica de evacuacion, mientras que
los mayores de 50 anos enfrentaron mayores dificultades fisicas
durante la evacuacion.

### 9e. Pasajeros de primera clase sobrevivieron mas que otros proporcionalmente

Creemos que si, los pasajeros de primera clase tuvieron mayor tasa de
supervivencia porque tenian cabinas en las cubiertas superiores, acceso
prioritario a los botes salvavidas y mayor capacidad economica para
obtener mejores ubicaciones en el barco.

In [23]:
# Tasa de supervivencia por clase de pasajero
supervivencia_clase = df_train.groupby('Pclass')['Survived'].agg(['sum', 'count', 'mean'])
supervivencia_clase.columns = ['Sobrevivieron', 'Total', 'Tasa']
supervivencia_clase['Tasa'] = (supervivencia_clase['Tasa'] * 100).round(2)

print("=== Supervivencia por clase ===")
print(supervivencia_clase.to_string())

=== Supervivencia por clase ===
        Sobrevivieron  Total   Tasa
Pclass                             
1                 136    216  62.96
2                  87    184  47.28
3                 119    491  24.24


### Observacion

Los datos confirman que los pasajeros de primera clase tuvieron una
tasa de supervivencia significativamente mayor que los de segunda y
tercera clase. Esto refleja las desigualdades sociales de la epoca
y el acceso diferenciado a los recursos de evacuacion.